src/IdempotentAPI/Filters/IdempotencyAttributeFilter.cs

```diff
@@ -72,54 +75,36 @@ public void OnActionExecuting(ActionExecutingContext context)
                     _distributedLockTimeout,
                     _cacheOnlySuccessResponses);
             }
-
-            _idempotency.ApplyPreIdempotency(context);
-        }
-
-        /// <summary>
-        /// Runs after the execution of the controller. In our case we used it to perform specific actions on Exceptions.
-        /// </summary>
-        /// <param name="context"></param>
-        public void OnActionExecuted(ActionExecutedContext context)
-        {
-            // If the Idempotency is disabled then stop.
-            // Stop if the PreIdempotency step is not applied.
-            if (!_enabled || _idempotency == null)
-            {
-                return;
-            }
-
-            if (context?.Exception is not null)
+            
+            await _idempotency.ApplyPreIdempotency(context);
+            
+            var result = await next();
+            if (result?.Exception is not null)
             {
-                _idempotency.CancelIdempotency();
+                await _idempotency.CancelIdempotency();
             }
```

Comment:
The `OnActionExecuted` applies the Pre-idempotency and, in conditions, returns the cached response based on the provided idempotencyKey. So, I was thinking of executing the `next()` only when the `_isPreIdempotencyCacheReturned` is False (i.e., when we do not have something to return by exposing it as read-only property). Otherwise, the following filter may alter our predefined cached response. Wdyt?

Reply:
Yes, that's it. Thank you for the information 😄